#02-CALIDAD Y LIMPIEZA

In [18]:
from logger_pipeline import registrar_transformacion
NOTEBOOK_2 = "02_calidad_y_limpieza.ipynb"
import pandas as pd
import pandas as pd
import numpy as np

df = pd.read_json(r"C:\Users\jorge leandro\OneDrive\Desktop\proyecto_final_md\data\raw\streaming_users_dirty.json")
streaming_users_clean = df.copy()

# Estandarización de variables catégoricas


In [19]:
# ESTANDARIZACIÓN DE 'subscription_plan'


#Ver el estado inicial 
print("=== ANTES: subscription_plan ===")
print(streaming_users_clean['subscription_plan'].value_counts(dropna=False))

#Quitar espacios invisibles y homogeneizar a formato "Nombre Propio"
#Esto unifica automáticamente: basico->Basico, BASICO->Basico, básico->Básico, estandar->Estandar, STANDARD->Standard, premium/PREMIUM->Premium
streaming_users_clean['subscription_plan'] = streaming_users_clean['subscription_plan'].astype(str).str.strip().str.title()

#Revertir strings 'None' o 'Nan' generados a valores nulos reales (NaN)
streaming_users_clean['subscription_plan'] = streaming_users_clean['subscription_plan'].replace({'None': np.nan, 'Nan': np.nan})

#Diccionario de mapeo específico para corregir abreviaturas, errores de ortografía e idiomas
mapeo_planes = {
    'Basico': 'Básico',     # Agrega la tilde faltante
    'Basic': 'Básico',      # Traduce del inglés
    'Std': 'Estándar',      # Expande la abreviatura
    'Estandar': 'Estándar',  # Agrega la tilde faltante
    'Standard': 'Estándar',  # Traduce del inglés
    'Premiun': 'Premium'     # Corrige el error de N por M
}

#Aplicar el mapa de correcciones
streaming_users_clean['subscription_plan'] = streaming_users_clean['subscription_plan'].replace(mapeo_planes)

#Verificar el resultado final agrupado
print("\n=== DESPUÉS: subscription_plan ===")
print(streaming_users_clean['subscription_plan'].value_counts(dropna=False))

=== ANTES: subscription_plan ===
subscription_plan
Básico       3450
Estándar     2711
Premium      1519
basico         60
BASICO         52
Basic          52
básico         50
Std            48
Estándar       46
estandar       36
STANDARD       34
Premium        31
PREMIUM        26
Premiun        23
premium        22
Name: count, dtype: int64

=== DESPUÉS: subscription_plan ===
subscription_plan
Básico      3664
Estándar    2875
Premium     1621
Name: count, dtype: int64


In [20]:
# ESTANDARIZACIÓN DE 'country'

#Ver el estado inicial 
print("=== ANTES: country ===")
print(streaming_users_clean['country'].value_counts(dropna=False))

#Quitar espacios y homogeneizar a formato "Nombre Propio"
# Esto unifica automáticamente las minúsculas: colombia/argentina/chile -> Colombia/Argentina/Chile
# También unifica los nombres en inglés que ya tienen formato correcto como Mexico o Peru
streaming_users_clean['country'] = streaming_users_clean['country'].astype(str).str.strip().str.title()

#Revertir strings 'None' o 'Nan' generados a valores nulos reales (NaN)
streaming_users_clean['country'] = streaming_users_clean['country'].replace({'None': np.nan, 'Nan': np.nan})

#Diccionario de mapeo específico para corregir abreviaturas, traducciones y tildes
mapeo_paises = {
    # Traducciones de idiomas
    'Brazil': 'Brasil',
    'Mexico': 'México',   # Asegura la tilde
    'Peru': 'Perú',       # Asegura la tilde
    
    # Abreviaturas (ISO 3166-1 alpha-3)
    'Arg': 'Argentina',
    'Bra': 'Brasil',
    'Chl': 'Chile',
    'Col': 'Colombia',
    'Mex': 'México',
    'Per': 'Perú',
    'Ury': 'Uruguay'
}

#Aplicar el mapa de correcciones
streaming_users_clean['country'] = streaming_users_clean['country'].replace(mapeo_paises)


=== ANTES: country ===
country
Brasil        1132
Chile         1132
México        1129
Uruguay       1124
Perú          1120
Colombia      1116
Argentina     1087
colombia        27
méxico          25
uruguay         24
Brazil          21
COL             19
CHL             18
URY             17
argentina       16
MEX             16
Chile           16
PER             16
Peru            15
BRA             15
chile           15
Mexico          15
brasil          13
perú            12
ARG             10
Argentina       10
Name: count, dtype: int64


In [21]:
#Verificar el resultado final agrupado
print("\n=== DESPUÉS: country ===")
print(streaming_users_clean['country'].value_counts(dropna=False))


=== DESPUÉS: country ===
country
México       1185
Chile        1181
Brasil       1181
Uruguay      1165
Perú         1163
Colombia     1162
Argentina    1123
Name: count, dtype: int64


In [22]:
# ESTANDARIZACIÓN DE 'favorite_genre'

#Ver el estado inicial 
print("=== ANTES: favorite_genre ===")
print(streaming_users_clean['favorite_genre'].value_counts(dropna=False))
#Quitar espacios en blanco invisibles y transformar a formato "Nombre Propio"
streaming_users_clean['favorite_genre'] = streaming_users_clean['favorite_genre'].astype(str).str.strip().str.title()

#Revertir strings 'None' o 'Nan' a nulos reales (NaN)
streaming_users_clean['favorite_genre'] = streaming_users_clean['favorite_genre'].replace({'None': np.nan, 'Nan': np.nan})

#Diccionario de mapeo específico para consolidar exactamente tus 7 categorías
mapeo_generos = {
    # Agrupación y traducción para Acción (absorbe "Accion" sin acento)
    'Accion': 'Acción',       
    'Action': 'Acción',       
    
    # Agrupación y traducción para Crimen y Comedia
    'Crime': 'Crimen',        
    'Comedy': 'Comedia',  
    
    # Agrupación y traducción para Documental (absorbe "Documentary" y la abreviatura)
    'Documentary': 'Documental',
    'Doc': 'Documental',      
    
    # Corrección de errores ortográficos de tipeo
    'Thriler': 'Thriller'     
}

#Aplicar el mapa de correcciones sobre la columna
streaming_users_clean['favorite_genre'] = streaming_users_clean['favorite_genre'].replace(mapeo_generos)


=== ANTES: favorite_genre ===
favorite_genre
Comedia        1112
Drama          1105
Documental     1095
Romance        1090
Thriller       1090
Acción         1082
Crime          1067
None            240
Action           22
COMEDIA          19
Crimen           18
Romance          17
CRIME            17
Documentary      16
DRAMA            16
Comedia          15
DOC              15
ROMANCE          14
ACCIÓN           14
THRILLER         14
comedy           12
Thriller         12
romance          12
documental       10
drama            10
accion            8
Drama             7
thriler           6
crime             5
Name: count, dtype: int64


In [23]:
# Verificar el resultado final perfectamente agrupado
print("=== DESPUÉS: favorite_genre ===")
print(streaming_users_clean['favorite_genre'].value_counts(dropna=False))

=== DESPUÉS: favorite_genre ===
favorite_genre
Comedia       1158
Drama         1138
Documental    1136
Romance       1133
Acción        1126
Thriller      1122
Crimen        1107
NaN            240
Name: count, dtype: int64


Se estandarizaron las columnas de plan, país y género favorito corrigiendo errores de ortografía, abreviaturas y mezclas de idiomas (inglés/español). De esta forma, garantizamos que los datos estén limpios, ordenados y listos para contarse correctamente sin duplicados visuales.

In [24]:

registrar_transformacion(NOTEBOOK_2, "Estandarización de variables categóricas (subscription_plan, country, favorite_genre)", df, streaming_users_clean)

# Tratamiento de registros duplicados
Duplicados en user_id 


In [11]:
# Conteo de filas con user_id repetido
total_id_duplicados = streaming_users_clean.duplicated(subset=['user_id']).sum()
print(f"Cantidad de filas con user_id repetido: {total_id_duplicados}")

Cantidad de filas con user_id repetido: 160


In [12]:
# DETECCIÓN Y VISUALIZACIÓN DE NULOS DISFRAZADOS
# Definimos los textos sospechosos
# Convertir nulos de texto a nulos reales de Pandas
nulos_disfrazados = ["NaN", "nan", "null", "NULL", "None", "none", "", " "]
streaming_users_clean=streaming_users_clean.replace(nulos_disfrazados, np.nan)

# Forzar la conversión de la fecha para destapar formatos corruptos
streaming_users_clean['last_login_date'] = pd.to_datetime(streaming_users_clean['last_login_date'], errors='coerce')

# CONSOLIDAR DUPLICADOS
# Ahora que Pandas "ve" los nulos reales, el groupby funcionará 

  # Definimos la función lógica que busca el último valor REAL y se salta los nulos
ultimo_valor_real = lambda x: x.dropna().iloc[-1] if not x.dropna().empty else np.nan

# Ejecutamos la consolidación inteligente
streaming_users_clean = streaming_users_clean.groupby('user_id').agg({
    'age': ultimo_valor_real,                      # Última edad real registrada
    'subscription_plan': ultimo_valor_real,         # Último plan real registrado
    'country': ultimo_valor_real,                   # Último país real registrado
    'last_login_date': ultimo_valor_real,           # Última fecha real registrada
    'favorite_genre': ultimo_valor_real,            # Último género real registrado
    'monthly_watch_time_mins': 'sum',               # Suma acumulada de minutos (ignora nulos)
    'customer_support_tickets': 'sum'               # Suma acumulada de tickets (ignora nulos)
}).reset_index()

# Verificación de control
print(f"Filas tras consolidar: {streaming_users_clean.shape[0]}")
print("\n=== CONTEO DE NULOS EN EL DATASET ===")
print(streaming_users_clean.isnull().sum())




Filas tras consolidar: 8000

=== CONTEO DE NULOS EN EL DATASET ===
user_id                       0
age                           0
subscription_plan             0
country                       0
last_login_date             759
favorite_genre              237
monthly_watch_time_mins       0
customer_support_tickets      0
dtype: int64


Nos aseguramos de destapar todos los nulos que venían escondidos como texto (como las palabras "null" o "NaN"). De lo contrario, el sistema se habría confundido y habría tomado esos textos falsos como datos reales.

Una vez que los nulos quedaron a la vista, agrupamos la información por usuario (user_id). Para lograrlo, usamos el siguiente criterio:

Para los datos personales (como país o edad), le pedimos al sistema que buscara el dato real más reciente y descartara los vacíos.

Para los datos de consumo (como los minutos vistos o los tickets de soporte), sumamos todo el historial del usuario.

Los valores nulos que aparecían en la columna de minutos vistos, se transformaron a '0' como consecuencia de la operación suma aplicada por Pandas en la agrupación 

Gracias a este orden, confirmamos que el dataset final tiene exactamente 8,000 usuarios únicos. Además, logramos detectar con precisión cuántos nulos reales quedan: 759 en la fecha de login y 237 en el género favorito. Lograr estos números limpios nos permite tomar mejores decisiones para rellenar esos vacíos sin romper ni inventar información.

In [253]:
#Conteo de duplicados totales
duplicados_totales = streaming_users_clean.duplicated().sum()
print(duplicados_totales)

0


Los duplicados puros (filas que son exactamente iguales en todas sus columnas) desaparecen por completo después del groupby debido a  la unificación por clave.

In [16]:

registrar_transformacion(NOTEBOOK_2, "Consolidación de duplicados por user_id", df, streaming_users_clean)

# Tratamiento de valores nulos
En columna categórica favorite_genre

In [254]:
# Rellenar con una etiqueta neutral que identifique la ausencia del dato
streaming_users_clean['favorite_genre'] = streaming_users_clean['favorite_genre'].fillna('Sin Especificar')

Para la variable cualitativa favorite_genre, se optó por un tratamiento que contempla un escenario MNAR (Missing Not at Random). Se asume que la ausencia de un género predilecto puede reflejar una conducta real del consumidor,y para evitar sesgar el perfil del usuario asignándole un gusto artificial mediante la moda general, se procedió a la creación de una categoría explícita denominada 'Sin Especificar'. Esto preserva la naturaleza del registro y permite segmentar analíticamente a los usuarios con perfiles indefinidos.

En columna last_login_date

In [255]:
# =====================================================================
# 1. IMPUTACIÓN BAJO SUPUESTO MNAR (Fecha fija para inactividad extrema)
# =====================================================================
fecha_penalizada = pd.to_datetime('2000-01-01')
streaming_users_clean['last_login_date'] = streaming_users_clean['last_login_date'].fillna(fecha_penalizada)



In [14]:

registrar_transformacion(NOTEBOOK_2, "Tratamiento de valores nulos", df, streaming_users_clean)

In [256]:
streaming_users_clean.head(50)

,user_id,age,subscription_plan,country,last_login_date,favorite_genre,monthly_watch_time_mins,customer_support_tickets
0,10000,39,Estándar,Brasil,2025-03-04,Crimen,805.8,99
1,10001,37,Estándar,Colombia,2019-04-02,Crimen,1173.4,2
2,10002,28,Básico,Colombia,2018-04-13,Crimen,401.0,0
3,10003,43,Básico,Uruguay,2021-01-31,Thriller,62.4,0
4,10004,51,Básico,Perú,2020-09-30,Thriller,477.8,1
5,10005,20,Básico,Uruguay,2020-07-03,Drama,670.2,2
6,10006,37,Básico,Perú,2019-07-26,Thriller,346.6,1
7,10007,31,Estándar,Chile,2019-02-24,Acción,974.6,1
8,10008,36,Premium,Colombia,2025-08-03,Romance,1432.2,2
9,10009,37,Estándar,Argentina,2024-02-12,Thriller,1375.4,1


Para la variable last_login_date, se rechaza el supuesto MCAR debido a que la ausencia de actividad suele estar ligada estadísticamente al comportamiento de abandono del cliente (inactividad). Al tratarse presumiblemente de un escenario MNAR (Missing Not at Random), la imputación mediante medidas de tendencia central común (como la media o mediana global) introduciría un sesgo optimista artificial. Por ello, se optó por una estrategia de imputación por valor arbitrario en el pasado, permitiendo preservar la señal de inactividad original para futuros modelos predictivos.

In [257]:
streaming_users_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8000 entries, 0 to 7999
Data columns (total 8 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   user_id                   8000 non-null   int64         
 1   age                       8000 non-null   int64         
 2   subscription_plan         8000 non-null   object        
 3   country                   8000 non-null   object        
 4   last_login_date           8000 non-null   datetime64[ns]
 5   favorite_genre            8000 non-null   object        
 6   monthly_watch_time_mins   8000 non-null   float64       
 7   customer_support_tickets  8000 non-null   int64         
dtypes: datetime64[ns](1), float64(1), int64(3), object(3)
memory usage: 500.1+ KB


Luego del tratamiento, se puede observar que no se encuentra existencia de valores faltantes

# Tratamiento de outliers
En columna age

In [258]:
# Definir los límites

limite_inferior = 13

limite_superior = 80

#Aplicar Capping Inferior (-5, 0 y 4 se convierten en 13)

streaming_users_clean['age'] = np.where(

streaming_users_clean['age'] < limite_inferior,

limite_inferior,

streaming_users_clean['age']

)

#Aplicar Capping Superior (130 y 150 se convierten en 80)

streaming_users_clean['age'] = np.where(

streaming_users_clean['age'] > limite_superior,

limite_superior,

streaming_users_clean['age']

)

#Comprobar el resultado para verificar las frecuencias

print("Distribución final de edades (extremos corregidos):")

print(streaming_users_clean['age'].value_counts().sort_index())  

Distribución final de edades (extremos corregidos):
age
13    450
14     70
15     83
16     85
17     92
     ... 
72      3
73      2
74      1
76      1
80     56
Name: count, Length: 64, dtype: int64


Tratamiento de edades mínimas imposibles (-5, 0 y 4 años): Se ajustaron a 13 años, alineando los datos con la edad mínima legal permitida en los términos de servicio de la aplicación. Esto permite conservar esos registros activos (evitando perder información valiosa de otras columnas) y mantiene el peso estadístico del segmento de usuarios más jóvenes, que es el más representativo de la plataforma.

Tratamiento de edades máximas imposibles (130 y 150 años): Al tratarse claramente de errores de registro o valores atípicos severos, se corrigieron estableciendo un límite máximo lógico de 80 años, saneando la distribución sin perder el rastro de los usuarios de mayor edad.

En columna monthly_watch_time_mins

In [259]:
# ==========================================
# 1. EXPLORACIÓN INICIAL & REDONDEO
# ==========================================

# Redondeamos los minutos a enteros (sin decimales)
streaming_users_clean['monthly_watch_time_mins'] = streaming_users_clean['monthly_watch_time_mins'].round(0)

# Ver distribución inicial para detectar outliers y errores brutos
print("Frecuencias iniciales (extremos):")
print(streaming_users_clean['monthly_watch_time_mins'].value_counts().sort_index())



# ==========================================
# 2. TRATAMIENTO DE ERRORES BRUTOS (Filtro Lógico)
# ==========================================

# Definimos una máscara para los datos que SÍ son lógicos
format_valido = (streaming_users_clean['monthly_watch_time_mins'] >= 0) & (streaming_users_clean['monthly_watch_time_mins'] < 50000)

# Calculamos la mediana usando solo los datos válidos
mediana_minutos = streaming_users_clean.loc[format_valido, 'monthly_watch_time_mins'].median()

# Reemplazamos los errores brutos (negativos y flags de sistema) con la mediana
streaming_users_clean['monthly_watch_time_mins'] = np.where(
    format_valido, 
    streaming_users_clean['monthly_watch_time_mins'], 
    mediana_minutos
)



# ==========================================
# 3. SUAVIZADO DE OUTLIERS (Capping por IQR)
# ==========================================

# Calculamos los límites estadísticos
Q1 = streaming_users_clean['monthly_watch_time_mins'].quantile(0.25)
Q3 = streaming_users_clean['monthly_watch_time_mins'].quantile(0.75)
IQR = Q3 - Q1
limite_superior = Q3 + 1.5 * IQR

# Aplicamos el recorte (Capping) en el límite superior
streaming_users_clean['monthly_watch_time_mins'] = np.where(
    streaming_users_clean['monthly_watch_time_mins'] > limite_superior, 
    limite_superior, 
    streaming_users_clean['monthly_watch_time_mins']
)



# ==========================================
# 4. VERIFICACIÓN FINAL
# ==========================================

print("\nTop 10 valores más altos tras la limpieza:")
print(streaming_users_clean['monthly_watch_time_mins'].value_counts().sort_index().tail(10))

Frecuencias iniciales (extremos):
monthly_watch_time_mins
-120.0        29
-1.0          19
 0.0         355
 2.0           1
 4.0           1
            ... 
 4450.0        1
 4546.0        1
 50000.0      11
 99999.0      19
 101313.0      1
Name: count, Length: 1704, dtype: int64

Top 10 valores más altos tras la limpieza:
monthly_watch_time_mins
1774.000      1
1777.000      1
1782.000      1
1815.000      2
1831.000      1
1834.000      1
1844.000      1
1845.000      1
1863.000      1
1891.625    182
Name: count, dtype: int64


Logramos limpiar con éxito la métrica de tiempo de visualización mensual mediante dos pasos clave:

Eliminación de errores: Reemplazamos los valores imposibles (como números negativos o códigos de error del sistema del tipo 99,999) por la duración promedio real de un usuario común.

Control de excesos: Ajustamos los consumos extremadamente altos a un límite máximo lógico para que no distorsionen los resultados finales.

En la columna customer_support_tickets

In [260]:


# ==========================================
# 1. EXPLORACIÓN INICIAL
# ==========================================

print("Distribución inicial de tickets (con errores):")
print(streaming_users_clean['customer_support_tickets'].value_counts().sort_index())
print("-" * 50)



# ==========================================
# 2. CÁLCULO DE LÍMITES ESTADÍSTICOS (IQR)
# ==========================================

# Filtramos temporalmente los códigos de error (ej. "99") para no sesgar los cuartiles
tickets_validos = streaming_users_clean.loc[streaming_users_clean['customer_support_tickets'] < 50, 'customer_support_tickets']

Q1 = tickets_validos.quantile(0.25)
Q3 = tickets_validos.quantile(0.75)
IQR = Q3 - Q1

# Calculamos los límites usando un factor de 3.0 para variables de conteo discreto
limite_inferior_iqr = Q1 - 1.5 * IQR
limite_superior_iqr = Q3 + 3.0 * IQR

# Ajustes lógicos y de formato (No hay tickets negativos y deben ser enteros)
limite_inferior = max(0, int(limite_inferior_iqr))
limite_superior = round(limite_superior_iqr)

print(f"Límite Inferior aplicado: {limite_inferior}")
print(f"Límite Superior aplicado: {limite_superior}")
print("-" * 50)



# ==========================================
# 3. APLICACIÓN DEL DOBLE CAPPING & FORMATO
# ==========================================

# np.clip limita los valores automáticamente entre el mínimo y máximo definido
streaming_users_clean['customer_support_tickets'] = np.clip(
    streaming_users_clean['customer_support_tickets'], 
    limite_inferior, 
    limite_superior
)

# Aseguramos el tipo de dato entero
streaming_users_clean['customer_support_tickets'] = streaming_users_clean['customer_support_tickets'].astype(int)



# ==========================================
# 4. VERIFICACIÓN FINAL
# ==========================================

print("Distribución final de Customer Support Tickets (Limpia):")
print(streaming_users_clean['customer_support_tickets'].value_counts().sort_index())

Distribución inicial de tickets (con errores):
customer_support_tickets
-1        28
 0      3564
 1      2754
 2      1200
 3       275
 4        91
 5        14
 6         5
 8         2
 99       35
 150      32
Name: count, dtype: int64
--------------------------------------------------
Límite Inferior aplicado: 0
Límite Superior aplicado: 4
--------------------------------------------------
Distribución final de Customer Support Tickets (Limpia):
customer_support_tickets
0    3592
1    2754
2    1200
3     275
4     179
Name: count, dtype: int64


Logramos normalizar el registro de tickets de soporte técnico mediante las siguientes acciones:

Aislamiento de errores: Ignoramos temporalmente los códigos de error (como el "99") para calcular los límites reales basados en el comportamiento común de los usuarios.

Ajuste de extremos (Capping): Fijamos el piso en 0 (ya que no existen tickets negativos) y limitamos los casos excesivos o erróneos a un techo máximo lógico (calculado estadísticamente).

En columna last_login_date

In [261]:
streaming_users_clean.loc[
    streaming_users_clean['last_login_date'] >= '2027-01-01',
    'last_login_date'
] = pd.Timestamp('2025-12-31')

In [15]:
# Ponelo justo después de redefinir streaming_users_clean con el .groupby()
registrar_transformacion(NOTEBOOK_2, "Tratamiento de outliers", df, streaming_users_clean)

In [262]:
streaming_users_clean['last_login_date'].value_counts()

last_login_date
2000-01-01    759
2025-12-31     18
2019-01-17     10
2025-09-29      9
2024-08-29      9
             ... 
2018-12-19      1
2024-08-24      1
2023-03-23      1
2023-02-10      1
2025-11-05      1
Name: count, Length: 2683, dtype: int64

In [263]:
streaming_users_clean.describe().round(2)

,user_id,age,last_login_date,monthly_watch_time_mins,customer_support_tickets
count,8000.00,8000.00,8000,8000.00,8000.00
mean,13999.50,33.85,2019-12-29 11:01:51.600000,763.95,0.84
min,10000.00,13.00,2000-01-01 00:00:00,0.00,0.00
25%,11999.75,25.00,2019-06-05 12:00:00,476.00,0.00
50%,13999.50,33.00,2021-09-09 00:00:00,752.00,1.00
75%,15999.25,42.00,2023-12-04 00:00:00,1042.25,1.00
max,17999.00,80.00,2025-12-31 00:00:00,1891.62,4.00
std,2309.55,12.20,NaN,415.01,0.95


In [264]:
streaming_users_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8000 entries, 0 to 7999
Data columns (total 8 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   user_id                   8000 non-null   int64         
 1   age                       8000 non-null   int64         
 2   subscription_plan         8000 non-null   object        
 3   country                   8000 non-null   object        
 4   last_login_date           8000 non-null   datetime64[ns]
 5   favorite_genre            8000 non-null   object        
 6   monthly_watch_time_mins   8000 non-null   float64       
 7   customer_support_tickets  8000 non-null   int64         
dtypes: datetime64[ns](1), float64(1), int64(3), object(3)
memory usage: 500.1+ KB


In [ ]:
# Exportamos la copia limpia a la carpeta de datos procesados
streaming_users_clean.to_csv(
    r"C:\Users\jorge leandro\OneDrive\Desktop\proyecto_final_md\data\processed\streaming_users_clean.csv",
    index=False)


In [17]:

registrar_transformacion(NOTEBOOK_2, "Exportación de dataset procesado", df, streaming_users_clean)